File: telegram-bot-source-jpa/src/main/java/io/github/dehuckakpyt/telegrambot/model/user/BaseTelegramUser.java
package io.github.dehuckakpyt.telegrambot.model.user;


import io.github.dehuckakpyt.telegrambot.model.UUIDTable;
import io.github.dehuckakpyt.telegrambot.model.source.TelegramUser;
import jakarta.persistence.Column;
import jakarta.persistence.MappedSuperclass;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.time.LocalDateTime;

/**
 * @author Denis Matytsin
 */
@MappedSuperclass
public class BaseTelegramUser extends UUIDTable implements TelegramUser {

    @Column(nullable = false, unique = true)
    private long userId;

    private String username;

    @Column(nullable = false)
    private String firstName;

    private String lastName;

    private String languageCode;

    @Column(nullable = false)
    private boolean available;

    @Column(nullable = false)
    private LocalDateTime updateDate;

    @Column(nullable = false)
    private LocalDateTime createDate;

    @Override
    public long getUserId() {
        return userId;
    }

    @Override
    public @Nullable String getUsername() {
        return username;
    }

    @Override
    public @NotNull String getFirstName() {
        return firstName;
    }

    @Override
    public @Nullable String getLastName() {
        return lastName;
    }

    @Override
    public @Nullable String getLanguageCode() {
        return languageCode;
    }

    @Override
    public boolean getAvailable() {
        return available;
    }

    @Override
    public @NotNull LocalDateTime getUpdateDate() {
        return updateDate;
    }

    @Override
    public @NotNull LocalDateTime getCreateDate() {
        return createDate;
    }

    public void setUserId(long userId) {
        this.userId = userId;
    }

    public void setUsername(@Nullable String username) {
        this.username = username;
    }

    public void setFirstName(@NotNull String firstName) {
        this.firstName = firstName;
    }

    public void setLastName(@Nullable String lastName) {
        this.lastName = lastName;
    }

    public void setLanguageCode(@Nullable String languageCode) {
        this.languageCode = languageCode;
    }

    public void setAvailable(boolean available) {
        this.available = available;
    }

    public void setUpdateDate(@NotNull LocalDateTime updateDate) {
        this.updateDate = updateDate;
    }

    public void setCreateDate(@NotNull LocalDateTime createDate) {
        this.createDate = createDate;
    }
}

File: telegram-bot-source-jpa/src/main/kotlin/io/github/dehuckakpyt/telegrambot/model/user/BaseTelegramUser.kt
package io.github.dehuckakpyt.telegrambot.model.user

import io.github.dehuckakpyt.telegrambot.model.UUIDTable
import io.github.dehuckakpyt.telegrambot.model.source.TelegramUser
import jakarta.persistence.Column
import jakarta.persistence.MappedSuperclass
import java.time.LocalDateTime


/**
 * Base realization with default fields. All fields will be filled automatically when user enter `/start`.
 *
 * I had to set zero, false, null values and use lateinit var.
 * I couldn't find another way to leave the instance creation to the empty constructor.
 * With an empty constructor, it is much easier to make convenient inheritance.
 *
 * @author Denis Matytsin
 */
@MappedSuperclass
abstract class BaseTelegramUser : UUIDTable(), TelegramUser {

    @Column(nullable = false, unique = true)
    override var userId: Long = 0

    override var username: String? = null

    @Column(nullable = false)
    override lateinit var firstName: String

    override var lastName: String? = null

    override var languageCode: String? = null

    @Column(nullable = false)
    override var available: Boolean = false

    @Column(nullable = false)
    override lateinit var updatedAt: LocalDateTime

    @Column(nullable = false)
    override lateinit var createdAt: LocalDateTime
}